dll/win32/shell32/dialogs/fprop.cpp

```diff
@@ -1,180 +1,310 @@
-    if (SUCCEEDED(hr))
+    static HRESULT ShowAsync(const CLSID *pClsidDefault, IDataObject *pDO,
+                             PFNINITIALIZE InitFunc, LPCWSTR InitString)
     {
-        pFileDefExt->AddRef(); // CreateInstance returns object with 0 ref count
-        hr = pFileDefExt->Initialize(HIDA_GetPIDLFolder(cida), pDataObj, NULL);
-        if (!FAILED_UNEXPECTEDLY(hr))
+        DATA *pData = (DATA*)SHAlloc(sizeof(*pData));
+        if (!pData)
+            return E_OUTOFMEMORY;
+        ZeroMemory(pData, sizeof(*pData));
+        pData->InitFunc = InitFunc;
+        if (InitString && FAILED(SHStrDupW(InitString, &pData->InitString)))
         {
-            hr = pFileDefExt->AddPages(AddPropSheetPageCallback, (LPARAM)&Header);
-            if (FAILED_UNEXPECTEDLY(hr))
-            {
-                ERR("AddPages failed\n");
-                return FALSE;
-            }
+            FreeData(pData);
+            return E_OUTOFMEMORY;
         }
-        else
+        if (pClsidDefault)
         {
-            ERR("Initialize failed\n");
-            return FALSE;
+            pData->ClsidDefault = *pClsidDefault;
+            pData->pClsidDefault = &pData->ClsidDefault;
         }
+
+        HRESULT hr = S_OK;
+        if (pDO)
+            hr = CoMarshalInterThreadInterfaceInStream(IID_IDataObject, pDO, &pData->pObjStream);
+
+        UINT flags = CTF_COINIT | CTF_PROCESS_REF | CTF_INSIST;
+        if (SUCCEEDED(hr) && !SHCreateThread(ShowPropertiesThread, pData, flags, NULL))
+        {
+            if (pData->pObjStream)
+                pData->pObjStream->Release();
+            hr = E_FAIL;
+        }
+        if (FAILED(hr))
+            FreeData(pData);
+        return hr;
     }
 
-    LoadPropSheetHandlers(wszPath, &Header, MAX_PROPERTY_SHEET_PAGE - 1, hpsxa, pDataObj);
+    static DWORD CALLBACK ShowPropertiesThread(LPVOID Param)
+    {
+        DATA *pData = (DATA*) Param;
+        CComPtr<IDataObject> pDO;
+        if (pData->pObjStream)
+            CoGetInterfaceAndReleaseStream(pData->pObjStream, IID_PPV_ARG(IDataObject, &pDO));
+        Show(pData->pClsidDefault, pDO, pData->InitFunc, pData->InitString);
+        FreeData(pData);
+        return 0;
+    }
+};
 
-    INT_PTR Result = PropertySheetW(&Header);
+static void CALLBACK
+FSFolderItemPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    // Add file-type specific pages
+    UNREFERENCED_PARAMETER(InitString);
+    CDataObjectHIDA cida(pDO);
+    if (SUCCEEDED(cida.hr()) && cida->cidl)
+    {
+        PCUITEMID_CHILD pidl = HIDA_GetPIDLItem(cida, 0);
+        AddFSClassKeysToArray(1, &pidl, hKeys, cKeys);
+    }
+}
 
-    for (UINT i = 0; i < 3; ++i)
-        if (hpsxa[i])
-            SHDestroyPropSheetExtArray(hpsxa[i]);
-    if (pFileDefExt)
-        pFileDefExt->Release();
+static void CALLBACK
+ProgIdPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
```

Comment:
Perhaps we could rename InitString to ProgId here (and in the rest of the PR)

Reply:
In this callback, perhaps.
These parameters are designed to be somewhat generic because it needs marshalling to the other thread. Then the callback is free to do what it wants with it.

Reply:
yeah but the callers even before the marshaling do know that this is still a progid, right?

Reply:
Of the 3 callers here, only one is really a ProgId.